c10/cuda/CUDAMallocAsyncAllocator.cpp

```diff
@@ -0,0 +1,823 @@
+  // static initializer?
+  if (pytorch_used_bytes[device] + size > pytorch_memory_limits[device]) {
+    err = cudaErrorMemoryAllocation;
+  } else {
+    err = cudaMallocAsync(devPtr, size, stream);
+  }
+
+  if (err == cudaErrorMemoryAllocation) {
+    // Clears CUDA's internal error state so the user, if desired, can catch the OOM
+    // exception, free some stuff on the script side, and retry the allocation.
+    // This aligns with the behavior of alloc_block in CUDACachingAllocator.cpp.
+    cudaGetLastError();
+    size_t device_free;
+    size_t device_total;
+    C10_CUDA_CHECK(cudaMemGetInfo(&device_free, &device_total));
+    TORCH_CHECK_WITH(CUDAOutOfMemoryError,
+                     false,
+                     "Allocation on device ", device, " would exceed allowed memory.",
+                     "\nCurrently allocated     : ", format_size(pytorch_used_bytes[device]),
+                     "\nRequested               : ", format_size(size),
+                     "\nDevice limit            : ", format_size(device_total),
+                     "\nFree (according to CUDA): ", format_size(device_free),
+                     "\nPytorch limit (set by user-supplied memory fraction)"
+                     "\n                        : ", format_size(pytorch_memory_limits[device]));
+  } else {
+    C10_CUDA_CHECK(err);
+  }
+
+  auto inserted = ptr_info.emplace(*devPtr, PtrUsage(size, capture_underway));
+  TORCH_INTERNAL_ASSERT(inserted.second,
+                        "address returned by cudaMallocAsync already exists "
+                        "in ptr_info");
+
+  inserted.first->second.usage_streams.emplace_back(stream, device);
+
+  pytorch_used_bytes[device] += size;
+}
+
+} // anonymous namespace
+
+// Same pattern as CUDACachingAllocator.cpp.
+// Again, I don't think we absolutely need the symmetry,
+// but it's the simplest way to imitate the interface.
+struct CudaCachingAllocator : public Allocator {
+  DataPtr allocate(size_t size) const override {
+    constexpr size_t one_exa_bytes = 1152921504606846976ULL;
+    TORCH_CHECK_WITH(
+        CUDAOutOfMemoryError,
+        size < one_exa_bytes,
+        "CUDA out of memory. Tried to allocate more than 1EB memory.");
+    int device;
+    C10_CUDA_CHECK(cudaGetDevice(&device));
+    void* r = nullptr;
+    if (size != 0) {
+      malloc(&r, device, size, cuda::getCurrentCUDAStream(device));
+    }
+    return {r, r, &raw_delete, Device(DeviceType::CUDA, device)};
+  }
+  DeleterFnPtr raw_deleter() const override {
+    return &raw_delete;
+  }
+};
+
+CudaCachingAllocator device_allocator;
+
+// Interface functions declared in CUDACachingAllocator.h
+
+Allocator* get(void) {
+  return &device_allocator;
+}
+
+// This function should not issue any context-creating calls,
+// just set up for later calls to init per-device pools based
+// on the current device each later call sees.
+void init(int dev_count) {
+  static bool called = false;
```

Comment:
As far as I can tell, you never set called to true. Also, you immediately take out a lock, so this doesn't need to be a static variable (which will cause its own synchronization code to be generated).

Reply:
> you never set called to true
oops lol

I reworked it, but maybe I need to be more careful generally about whether init is called before any of the allocator's other functions. Are there external guarantees init will be called before any of the other functions?

Reply:
I don't remember, I think we saw some weird errors (now fixed) when we moved to per-device allocator because some functions were called before init, but I don't remember the details now